platform/USBPhy.h

```diff
@@ -0,0 +1,380 @@
+    virtual void init(USBPhyEvents *events) = 0;
+
+    /**
+     * Power down this USBPhy instance
+     *
+     * Disable interrupts and stop sending events.
+     */
+    virtual void deinit() = 0;
+
+    /**
+     * Make the USB phy visible to the USB host
+     *
+     * Enable either the D+ or D-  pullup so the host can detect
+     * the presence of this device.
+     */
+    virtual void connect(void) = 0;
+
+    /**
+     * Detach the USB phy
+     *
+     * Disable the D+ and D- pullup and stop responding to
+     * USB traffic.
+     */
+    virtual void disconnect(void) = 0;
+
+    /**
+     * Set this device to the configured state
+     *
+     * Enable added endpoints if they are not enabled
+     * already.
+     */
+    virtual void configure(void) = 0;
+
+    /**
+     * Leave the configured state
+     *
+     * This is a notification to the USBPhy indicating that the device
+     * is leaving the configured state. The USBPhy can disable all
+     * endpoints other than endpoint 0.
+     *
+     */
+    virtual void unconfigure(void) = 0;
+
+    /**
+     * Enable the start of frame interrupt
+     *
+     * Call USBPhyEvents::sof on every frame.
+     */
+    virtual void sof_enable() = 0;
+
+    /**
+     * Disable the start of frame interrupt
+     *
+     * Stop calling USBPhyEvents::sof.
+     */
+    virtual void sof_disable() = 0;
+
+    /**
+     * Set the USBPhy's address
+     *
+     * @param address This device's USB address
+     */
+    virtual void set_address(uint8_t address) = 0;
+
+    /**
+     * Wake upstream devices
+     */
+    virtual void remote_wakeup(void) = 0;
+
+    /**
+     * Get the endpoint table
+     *
+     * This function returns a table which describes the endpoints
+     * can be used, the functionality of those endpoints and the
+     * resource cost.
+     */
+    virtual const usb_ep_table_t* endpoint_table() = 0;
+
+    /**
+     * Set wMaxPacketSize of endpoint 0
+     *
+     * @param max_packet The wMaxPacketSize value for endpoint 0
+     * @return The actual size of endpoint 0
+     */
+    virtual uint32_t ep0_set_max_packet(uint32_t max_packet) = 0;
+
+    /**
+     * Read the contents of the SETUP packet
+     *
+     * @param buffer Buffer to fill with data
+     * @param size Size of buffer passed in
+     */
+    virtual void ep0_setup_read_result(uint8_t *buffer, uint32_t size) = 0;
+
+    /**
+     * Start receiving a packet of up to wMaxPacketSize on endpoint 0
+     */
+    virtual void ep0_read(void) = 0;
+
+    /**
+     * Read the contents of a received packet
+     *
+     * @param buffer Buffer to fill with the data read
+     * @param size Size of buffer
+     */
+    virtual uint32_t ep0_read_result(uint8_t *buffer, uint32_t size) = 0;
+
+    /**
+     * Write a packet on endpoint 0
+     *
+     * @param buffer Buffer fill with data to send
+     * @param size Size of data to send
+     */
+    virtual void ep0_write(uint8_t *buffer, uint32_t size) = 0;
+
```

Comment:
Should we check for buffer overflows or other such conditions? This header may not be the right place but I couldn't find any checks for that so far..

Reply:
This is just the interface class. Its up to the vendors implementation if they want to check if the buffer is too large. By design USBDevice limits the size this function is called with to the max packet size for endpoint 0, so a check by the vendor code will be somewhat redundant, but doesn't hurt.